go/vt/binlog/keyrange_filter.go

```diff
@@ -60,6 +78,56 @@ func KeyRangeFilterFunc(keyrange *topodatapb.KeyRange, sendReply sendTransaction
 	}
 }
 
+func getValidRangeQuery(sql string, keyspaceIDs [][]byte, keyrange *topodatapb.KeyRange) (query string, err error) {
+	statement, err := sqlparser.Parse(sql)
+	_, trailingComments := sqlparser.SplitTrailingComments(sql)
+	if err != nil {
+		return "", err
+	}
+
+	switch statement := statement.(type) {
+	case *sqlparser.Insert:
+		query, err := generateSingleInsertQuery(statement, keyspaceIDs, trailingComments, keyrange)
+		if err != nil {
+			return "", err
+		}
+		return query, nil
+	default:
+		return "", errors.New("unsupported construct ")
+	}
+}
+
+func generateSingleInsertQuery(ins *sqlparser.Insert, keyspaceIDs [][]byte, trailingComments string, keyrange *topodatapb.KeyRange) (query string, err error) {
+	var values sqlparser.Values
+	switch rows := ins.Rows.(type) {
+	case *sqlparser.Select, *sqlparser.Union:
+		return "", errors.New("unsupported: insert into select")
+	case sqlparser.Values:
+		values = rows
+		if len(values) != len(keyspaceIDs) {
+			return "", fmt.Errorf("length of values tuples %v doesn't match with length of keyspaceids %v", len(values), len(keyspaceIDs))
+		}
+		queryBuf := sqlparser.NewTrackedBuffer(nil)
+		queryBuf.Myprintf("insert %v%sinto %v%v values", ins.Comments, ins.Ignore, ins.Table, ins.Columns)
+		prefix := ""
+		for rowNum, val := range values {
+			if key.KeyRangeContains(keyrange, keyspaceIDs[rowNum]) {
+				queryBuf.Myprintf("%s%v", prefix, val)
+				prefix = ", "
+			}
+		}
+		if prefix == "" {
+			return "", nil
+		}
+		queryBuf.Myprintf("%v%s", ins.OnDup, trailingComments)
+		query := queryBuf.String()
+		return query, nil
```

Comment:
In this section, could we just build a new 'rows' value, and then re-print the SQL statement as is with it, instead of re-generating it all? Would that be simpler?